### ISSUE TITLE
Chapter 6 - page 150 'post_new_course_db' function is failing to compile

### ISSUE THREAD

**Issue by RPG-Alex** [peshwar9/rust-servers-services-apps#6]
In your book and in your source code here you have an issue of the compiler being able to infer the types between Course and NewCourse, which makes sense given these two structs are different enough (this is in the iter5/handlers/course.rs file for reference. You have this code:
```
pub async fn post_new_course(
    new_course: web::Json<CreateCourse>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse, EzyTutorError> {
    post_new_course_db(&app_state.db, new_course.into()?)
        .await
        .map(|course| HttpResponse::Ok().json(course))
}
```
Which results in this error:
```
error[E0283]: type annotations needed
  --> src/bin/../iter5/handlers/course.rs:48:50
   |
48 |     post_new_course_db(&app_state.db, new_course.into()?)
   |                                                  ^^^^
   |
   = note: cannot satisfy `_: std::convert::From<actix_web::web::Json<models::course::CreateCourse>>`
   = note: required for `actix_web::web::Json<models::course::CreateCourse>` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
   |
48 |     post_new_course_db(&app_state.db, <actix_web::web::Json<models::course::CreateCourse> as Into<T>>::into(new_course)?)
   |                                       ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++          ~

For more information about this error, try `rustc --explain E0283`.
error: could not compile `chapter-6` (bin "iter5") due to 1 previous error
```

Im not really sure how you got it to compile when you wrote it that way. I cannot and I have cloned your chapter six codebase and am still unable to get compilation.

My solution is to directly grab the first value from the tuple like so:
```
use crate::dbaccess::course::*;
use crate::errors::EzyTutorError;
use crate::models::course::{CreateCourse, Course, UpdateCourse};
use crate::state::AppState;
use actix_web::{web, HttpResponse};

// rest of the code

pub async fn post_new_course(
    new_course: web::Json<CreateCourse>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse, EzyTutorError> {
    post_new_course_db(&app_state.db, new_course.0)
        .await
        .map(|course| HttpResponse::Ok().json(course))
}
```
This compiles and works.

**Comment by koalabi** [comment]
Thank you for reporting this problem and suggesting a fix.

No, for your information, all the code was verified at the time of finalizing the book and, at the time, everything compiled fine Maybe the compiler has evolved since then or some of the crate dependencies ...

**Comment by RPG-Alex** [comment]
> Thank you for reporting this problem and suggesting a fix.
> 
> No, for your information, all the code was verified at the time of finalizing the book and, at the time, everything compiled fine Maybe the compiler has evolved since then or some of the crate dependencies ...

Thank you for responding so quickly! I am enjoying the book so far, and it has been illuminating for rust for servers. I am looking forward to the p2p chapter (its most of the reason I bought the book actually). [Cargo version locks I thought](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)?, and I used your toml so I thought the crate dependencies would not be able to cause the issue?

Regarding the compiler, you mean the version of Rustc? Sure I'll share my version here, I am using nightly so it may be possible:
`rustc 1.77.0-nightly (b6a8c762e 2024-01-06)`

Though switching to the most current stable version is yielding the same issue. So I am not sure when that inference became an issue for the compiler. Thanks for the response. Its working now.

**Comment by koalabi** [comment]
I will have a look, just to try to understand. (I'm not the author; I just helped finalizing the source code - we had to upgrade the compiler and some crates at that time). I don't remember whether the version constraints are strict ("this" version number only or more relaxed "this version number or later"). If they are fixed, I would tend to think the same as you, namely crates shouldn't be the problem ... For the compiler version, I have to go back and try to retrieve the compiler version. Thanks for your feedback anyway.
